### ISSUE TITLE
`pull_request_target is almost always used insecurely`

### ISSUE THREAD

**Issue by ljharb** [woodruffw/zizmor#82]
pull_request_target is more secure than pull_request, which is why it gets access to secrets. There's nothing "fundamentally insecure" about it.

Can you elaborate on this?

**Comment by woodruffw** [comment]
Hi @ljharb, thanks for filing an issue.

The docs (<https://woodruffw.github.io/zizmor/audits/#dangerous-triggers>) have some sources for this, but to summarize:

* `pull_request_target` (and `workflow_call`) are somewhat unique in that they're often triggered by third-party users, but have the same *context* as their upstream repo. That means that they have secret and other context access for the upstream repo, rather than the repo sending the PR. There are other triggers that run in "upstream" contexts, but most can only be triggered by that upstream itself (e.g. `push`, since pushing to a repo already implies a degree of access to it).
* *On its own*, this is not an issue. However, users *consistently* misuse `pull_request_target` (and `workflow_call`) in workflows in ways that allow someone to achieve code execution *in the upstream context* just by sending a PR.

The "classic" example of this is documented in a [2021 post from GitHub's security lab](https://securitylab.github.com/resources/github-actions-preventing-pwn-requests/):

```yaml
on:
  pull_request_target

jobs:
  build:
    name: Build and test
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
      with:
        ref: ${{ github.event.pull_request.head.sha }}

    - uses: actions/setup-node@v1
    - run: |
        npm install
        npm build
```

The upstream's *intent* here is (typically) just to install some dependencies. However, because they've checked out attacker-controlled code into the upstream context, `npm install` is fully controlled by the attacker instead, and can exfiltrate the default `GH_TOKEN`, any members of `secrets.*`, etc.

TL;DR: `pull_request_target` runs in the parent repo context but, unlike most parent repo triggers, can be triggered by external users as well. This enables footguns around attacker-controlled code in the parent context that alternative triggers (like `pull_request`) don't allow.

(To my understanding, GitHub is aware of this footgun and considers it a security issue, per the GHSL post above and [warnings in the docs](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#pull_request_target). However, they aren't interested in removing or neutralizing `pull_request_target` because they consider it the user's responsibility to use it correctly. Hence `zizmor` flagging on it.)

**Comment by ljharb** [comment]
How can i override that warning, then, given that the *entire* point of pull_request_target is to have access to secrets, and that i am using it securely?

**Comment by woodruffw** [comment]
> How can i override that warning, then, given that the _entire_ point of pull_request_target is to have access to secrets, and that i am using it securely?

You can't yet, at least not directly within `zizmor` itself -- I need to add support for some kind of ignoring mechanism per #12. If you're using the SARIF integration, then you can use GitHub's UI to dismiss the finding, however.

Out of curiosity, can you share the workflow that you think is using `pull_request_target` securely? IME with security reviews, anything that uses it beyond its most locked-down case (no code execution, just used to send comments/change issues) tends to get something wrong with it.

Longer term, it's possible that I could refine the dangerous-triggers audit to be more precise about `pull_request_target`. The reason that hasn't happened yet is because I'm currently erring more towards false positives than false negatives, since this was originally designed primarily as an auditing assistant.

**Comment by ljharb** [comment]
Email, Twitter, whatever you like - I’m ljharb on everything.

**Comment by ljharb** [comment]
The issue should probably still be open, though, because I need a way to not warn on the allow edits workflow, at least.

**Comment by ljharb** [comment]
Sure!

**Comment by QuLogic** [comment]
> Thanks for the links. `require-allow-edits.yml` looks fine (and is a good example of how `pull_request_target` _can_ be used safely),

But can `zizmor` be changed to detect these sorts of workflows? Ones without substitutions or user-induced input should be safe (IIUC).

I don't want to ignore the check; I want `zizmor` to tell me concretely if it becomes insecure.

**Comment by woodruffw** [comment]
(Sorry for the wall of text that follows.)

> But can `zizmor` be changed to detect these sorts of workflows?

In the long term yes, but not at the moment. There are two big limitations (which I'm working on/thinking about how to accomplish) that prevent this currently:

1. All audits are currently independent of each other, i.e. `dangerous-triggers` doesn't know about `template-injection` or vice versa. Long term audits should be able to declare dependency relationships on each other, and use a previous audit's results to decide whether or not to emit findings of their own.
2. `zizmor` currently audits only *workflows*, not the contents of the *actions* they call. Once we're able to do the latter, (1) will compose nicely so that we can express things like "only warn about a dangerous trigger if the workflow itself looks fine *and* every action it calls looks fine."

These are pretty long term goals, however, and I'm not sure if there's an actual stable generality here yet -- many of `zizmor`'s analyses are imprecise because they have to jump between YAML, bash, JS, etc., which means that I've erred on the site of emitting false positives for "very hard to use correctly" things like `pull_request_target`.

> Ones without substitutions or user-induced input should be safe (IIUC).

For the most part, although there are caveats here:

1. Any `pull_request_target` or similar that modifies the environment via `GITHUB_ENV` is susceptible to code execution, if the attacker can manipulate the environment variable names/values to some degree.
2. It's hard to describe "input" to a given workflow with full generality, because of how dynamic they are at runtime. For example: a `pull_request_target` might not "take" any input, but it might pull data from an upstream *or* a fork based on some event state. That data won't show up as an explicit `${{ some.context }}` reference, but it could be attacker-controllable in dangerous ways. GHSA-5xr6-xhww-33m4 is *kind of* an example of this (not exactly since it's a different kind of ambiguity, but it demonstrates how GitHub's internal "provenance" of information diverges from how users think about user/repo/workflow identities).
3. More broadly, public understanding of GHA's attack surface is still pretty young -- template injections are well understood, but the dangers of `GITHUB_ENV` weren't super well understood until last year and cache key exfil/poisoning is new to this year. So more generally, I don't feel comfortable asserting that `zizmor` *knows* that a given `pull_request_target`-driven workflow *is* safe, given that people keep finding new ways to punch holes in them :slightly_smiling_face: 

TL;DR: I have plans to make this more precise, but it's not trivial!
